### PR TITLE
Eyes Universal

### PR DESCRIPTION
**Pull request by kyryloonufriiev**

### CODE REVIEW

File: packages/eyes-universal/SPEC.md [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+
+## WebSocket
+The client-server architecture of the universal sdk requires an implementation of the communication layer between the *Client* and the *Server*. Because of a major need for bidirectional communication, WebSocket protocol was chosen. WebSocket protocol operates only with messages, and each of those is an independent chunk of data. The protocol doesn't support getting a response on a message. This is why the format of communication is determined by a proprietary request-response messaging interface (*Universal SDK messaging protocol*), which requires a specific format of client-server messages.
+
+All of the commands should be treated as requests, which means that response is always has to be sent after a request is received. But simple events are also allowed by the *Universal SDK messaging protocol*, which means that *Client*, as well as *Server*, could send a message which doesn't require any response.
+
+Reference implementation: [JS implementation](https://github.com/applitools/eyes.sdk.javascript1/blob/1a54c6b3f28b25b41a708f5b600ceabbf8fc9db6/packages/eyes-universal/src/socket.js)
```
Comment (amitzur) [inline packages/eyes-universal/SPEC.md:15]: add ruby implementation reference?

File: packages/eyes-universal/SPEC.md [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+```ts
+{
+  name: string, // could be any string, but recommended format is "<Domain>.<eventName>"
+  payload?: any // any event data
+}
+```
+### Client-initiated events
+In order to pass some data to the *Server* and let *Server* process it on its own *Client* could send lightweight events ([Event format](#Event-format)), which will not be responded in any way.
+
+#### Session.init
```
Comment (amitzur) [inline packages/eyes-universal/SPEC.md:57]: add reference example

File: packages/eyes-universal/SPEC.md [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+
+### Client-initiated commands
+In order to perform any action, the *Client* has to send a proper request to the *Server* in a specific format ([Request format](#Request-format)) and wait for the response in a format described here ([Response format](#Response-format)).
+
+#### EyesRunner.new
+This request should be sent to create a runner object. It expects input of type [RunnerConfiguration](https://github.com/applitools/eyes.sdk.javascript1/blob/63e2f6fd3a1751c0ac2edce5971391aad0a2fd78/packages/eyes-api/src/Runners.ts#L6-L10).
+
+In response client should expect to get a runner reference, this reference has to be used in runner related requests ([EyesRunner.open](####EyesRunner.open), [EyesRunner.close](####EyesRunner.close))
+
+> Do not send this command in a moment when `EyesRunner` is constructed but instead send it lazily when the actual eyes object has to be opened. Pay attention that in this architecture eyes could be created only from a runner instance, and creation and opening of the eyes are combined in one operation.
```
Comment (amitzur) [inline packages/eyes-universal/SPEC.md:75]: why does that matter?

File: packages/eyes-universal/SPEC.md [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+
+#### EyesRunner.open
```
Comment (amitzur) [inline packages/eyes-universal/SPEC.md:77]: example of creating `DriverRef`. Is it different from creating the `RunnerRef`?

File: packages/eyes-universal/SPEC.md [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+This command is meant to be used to close all eyes objects created with this runner and return results from each of the eyes objects. It doesn't expect any input except a related RunnerRef (from [EyesRunner.new](####EyesRunner.new):
+```ts
+{
+    runner: RunnerRef,
+}
+```
+
+In response client will receive an array of [TestResultsContainer](https://github.com/applitools/eyes.sdk.javascript1/blob/63e2f6fd3a1751c0ac2edce5971391aad0a2fd78/packages/eyes-api/src/output/TestResultsContainer.ts#L3-L6)'s
+
+> There is no way to throw on diffs or new tests, this command will never throw an error in such cases.
```
Comment (amitzur) [inline packages/eyes-universal/SPEC.md:99]: there is no way? Is throwing errors even part of the protocol? Or do you mean return `error` field in the response?

File: packages/eyes-universal/SPEC.md [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+This command has to be used in order to close eyes object and finish the test. It doesn't expect any input except a related EyesRef.
+```ts
+{
+    eyes: EyesRef,
+}
+```
+
+In a case of success, the client will receive a response with [TestResults](https://github.com/applitools/eyes.sdk.javascript1/blob/63e2f6fd3a1751c0ac2edce5971391aad0a2fd78/packages/eyes-api/src/output/TestResults.ts#L16-L47) object.
+
+> There is no way to throw on diffs or new tests, this command will never throw an error in such cases.
```
Comment (amitzur) [inline packages/eyes-universal/SPEC.md:171]: same question as line 99: there is no way? Is throwing errors even part of the protocol? Or do you mean return `error` field in the response?

**Comment by amitzur** [comment]
need to add details about cli to spawn process

**Comment by kyryloonufriiev** [comment]
This is my proposition for internal API, to make JS Core and Universal SDK symmetrical.

```ts
interface Core {
  makeClient(options: {type: 'vg' | 'classic', concurrency?: number, legacy?: boolean}): Promise<Client>
  getViewportSize(options: {driver: Driver}): Promise<{width: number,  height: number}>
  setViewportSize(options: {driver: Driver, viewportSize: {width: number,  height: number}}): Promise<void>
  closeBatches(options: any): Promise<void>
  deleteTest(options: any): Promise<void>
}

interface Client {
  makeEyes(options: {driver: DriverRef, config: Configuration}): Promise<Eyes>
  getResults(): Promise<TestResultsSummary>
}

interface Eyes {
  check(options: {settings: CheckSettings, config: Configuration}): Promise<CheckResult>
  locate(options: {settings: VisualLocatorSettings, config: Configuration}): Promise<{[key: string]: Region[]}>
  extractTextRegions(options: {settings: OCRSettings, config: Configuration}): Promise<{[key: string]: string[]}>
  extractText(options: {regions: OCRRegions[], config: Configuration}): Promise<string[]>
  close(): Promise<TestResults>
  abort(): Promise<TestResults>
}

interface CoreWS {
  on('Core.makeClient', callback: (options: {type: 'vg' | 'classic', concurrency?: number, legacy?: boolean}) => Promise<ClientRef>)
  on('Core.getViewportSize', callback: (options: {driver: DriverRef}) => Promise<{width: number,  height: number}>)
  on('Core.setViewportSize', callback: (options: {driver: DriverRef, viewportSize: {width: number,  height: number}}) => Promise<void>)
  on('Core.closeBatches', callback: (options: any) => Promise<void>)
  on('Core.deleteTest', callback: (options: any) => Promise<void>)

  on('Client.makeEyes', callback: (options: {client: ClientRef, driver: DriverRef, config: Configiration}) => Promise<EyesRef>)
  on('Client.getResults', callback: (options: {client: ClientRef}) => Promise<TestResultsSummary>)

  on('Eyes.check', callback: (options: {eyes: EyesRef, settings: CheckSettings, config: Configuration}) => Promise<CheckResult>)
  on('Eyes.locate', callback: (options: {eyes: EyesRef, settings: VisualLocatorSettings, config: Configuration}) => Promise<{[key: string]: Region[]}>)
  on('Eyes.extractTextRegions', callback: (options: {eyes: EyesRef, settings: OCRSettings, config: Configuration}) => Promise<{[key: string]: string[]}>)
  on('Eyes.extractText', callback: (options: {eyes: EyesRef, regions: OCRRegions[], config: Configuration}) => Promise<string[]>)
  on('Eyes.close', callback: (options: {eyes: EyesRef}) => Promise<TestResults>)
  on('Eyes.abort', callback: (options: {eyes: EyesRef}) => Promise<TestResults>)
}

```

**Pull request merged**